zrml/swaps/src/lib.rs

```diff
@@ -1365,75 +1365,85 @@ mod pallet {
                 .fold(false, |acc, (&x, &y)| acc || x == y);
             ensure!(!has_duplicates, Error::<T>::SomeIdenticalAssets);
 
-            if scoring_rule == ScoringRule::CPMM {
-                ensure!(amount.is_some(), Error::<T>::InvalidAmountArgument);
-                ensure!(
-                    amount_unwrapped >= T::MinLiquidity::get(),
-                    Error::<T>::InsufficientLiquidity
-                );
-                let _ = swap_fee.ok_or(Error::<T>::InvalidFeeArgument)?;
-                let weights_unwrapped = weights.ok_or(Error::<T>::InvalidWeightArgument)?;
-                Self::check_provided_values_len_must_equal_assets_len(&assets, &weights_unwrapped)?;
-
-                for (asset, weight) in assets.iter().copied().zip(weights_unwrapped) {
-                    let free_balance = T::Shares::free_balance(asset, &who);
-                    ensure!(free_balance >= amount_unwrapped, Error::<T>::InsufficientBalance);
-                    ensure!(weight >= T::MinWeight::get(), Error::<T>::BelowMinimumWeight);
-                    ensure!(weight <= T::MaxWeight::get(), Error::<T>::AboveMaximumWeight);
-                    map.insert(asset, weight);
-                    total_weight = total_weight.check_add_rslt(&weight)?;
-                    T::Shares::transfer(asset, &who, &pool_account, amount_unwrapped)?;
-                }
+            let (pool_status, total_subsidy, total_weight, weights, pool_amount) =
+                match scoring_rule {
+                    ScoringRule::CPMM => {
+                        ensure!(amount.is_some(), Error::<T>::InvalidAmountArgument);
+                        ensure!(
+                            amount_unwrapped >= T::MinLiquidity::get(),
+                            Error::<T>::InsufficientLiquidity
+                        );
+                        swap_fee.ok_or(Error::<T>::InvalidFeeArgument)?;
```

Comment:
You could also got with `ensure!(swap_fee.is_some())` here. But this is nitpick territory. ;)

Reply:
No, I think it's a really good suggestion, because `ok_or` unwraps the value, which is not needed here.